### PR TITLE
[NNC] Implementation for aten::cat without conditionals.

### PR DESCRIPTION
**Pull request by navahgar**
This PR adds an implementation for `aten::cat` in NNC without any conditionals. This version is not enabled by default.

Here is the performance of some micro benchmarks with and without conditionals. There is up to 50% improvement in performance without conditionals for some of the shapes. 

aten::cat implementation in NNC **with** conditionals
```
$ python -m benchmarks.tensorexpr --device cpu --mode fwd --jit_mode trace --cpu_fusion concat
pt: concat2d2input_fwd_cpu_1_160_1_14_1: 5.44 us, SOL 0.26 GB/s, algorithmic 0.51 GB/s
pt: concat2d2input_fwd_cpu_1_580_1_174_1: 5.75 us, SOL 1.05 GB/s, algorithmic 2.10 GB/s
pt: concat2d2input_fwd_cpu_20_160_20_14_1: 6.87 us, SOL 4.05 GB/s, algorithmic 8.11 GB/s
pt: concat2d2input_fwd_cpu_20_580_20_174_1: 14.52 us, SOL 8.31 GB/s, algorithmic 16.62 GB/s
pt: concat2d2input_fwd_cpu_8_512_8_512_1: 9.58 us, SOL 6.84 GB/s, algorithmic 13.68 GB/s
```
aten::cat implementation in NNC **without** conditionals
```
$ python -m benchmarks.tensorexpr --device cpu --mode fwd --jit_mode trace --cpu_fusion --cat_wo_conditionals concat
pt: concat2d2input_fwd_cpu_1_160_1_14_1: 4.67 us, SOL 0.30 GB/s, algorithmic 0.60 GB/s
pt: concat2d2input_fwd_cpu_1_580_1_174_1: 5.65 us, SOL 1.07 GB/s, algorithmic 2.14 GB/s
pt: concat2d2input_fwd_cpu_20_160_20_14_1: 6.10 us, SOL 4.56 GB/s, algorithmic 9.12 GB/s
pt: concat2d2input_fwd_cpu_20_580_20_174_1: 7.44 us, SOL 16.22 GB/s, algorithmic 32.44 GB/s
pt: concat2d2input_fwd_cpu_8_512_8_512_1: 6.46 us, SOL 10.14 GB/s, algorithmic 20.29 GB/s
```

### CODE REVIEW

**Review by bertmaher — APPROVED** [review]

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 1]
```diff
@@ -1995,6 +2003,111 @@ Tensor* TensorExprKernel::computeSoftmax(
...
+  // for i
+  //   for j
+  //     for k
+  //       output[i,j+l2,k] = inp3[i,j,k]
+
+  auto output_sizes = inferSizesForValue(v);
+  std::vector<const Expr*> output_sizes_expr(output_sizes.size());
+  for (size_t i = 0; i < output_sizes.size(); ++i) {
+    output_sizes_expr[i] = output_sizes[i].node();
+  }
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/kernel.cpp:2060]: Nit: you probably could do
```
std::vector<const Expr*> output_sizes_expr = ExprHandleVectorToExprVector(output_sizes);
```

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 2]
```diff
@@ -1995,6 +2003,111 @@ Tensor* TensorExprKernel::computeSoftmax(
...
+  auto output_buf = new Buf(
+      new Var("aten_cat", kHandle), output_sizes_expr, ToDtype(highType));
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/kernel.cpp:2062]: Nit: No need to create `Var` directly, you could pass a name to a `Buf`'s constructor directly:
```
auto output_buf = new Buf("aten_cat", output_sizes_expr, ToDtype(highType));
```

File: torch/_C/__init__.pyi.in [hunk 3]
```diff
@@ -208,6 +208,7 @@ def _jit_override_can_fuse_on_gpu(override: _bool): ...
 def _jit_set_texpr_fuser_enabled(enable: _bool): ...
 def _jit_set_te_must_use_llvm_cpu(use_llvm: _bool): ...
 def _jit_set_nvfuser_enabled(enable: _bool) -> _bool: ...
+def _jit_cat_wo_conditionals(optimize_cat: _bool): ...
```
Comment (ZolotukhinM) [inline torch/_C/__init__.pyi.in:211]: Isn't this file autogenerated? I don't remember ever touching it manually

**Review by ZolotukhinM — COMMENTED** [review]
Sorry, if I'm too late for the party! I think this change needs some tests with it (not just benchmarks), and I also have a couple of minor remarks.

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 1 above)
Comment (navahgar) [reply torch/csrc/jit/tensorexpr/kernel.cpp]: Done.

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 2 above)
Comment (navahgar) [reply torch/csrc/jit/tensorexpr/kernel.cpp]: Done.

File: torch/_C/__init__.pyi.in (hunk 3 above)
Comment (navahgar) [reply torch/_C/__init__.pyi.in:211]: Nope. There are lots of commits to this file. There is another file `__init__.pyi` which is auto-generated I think.

**Comment by navahgar** [comment]
Added a unit test as well. @ZolotukhinM

**Pull request closed without merging**